### PR TITLE
Implement toLowerCase and toUpperCase built-in functions.

### PR DESCRIPTION
**Pull request by zherczeg**
JerryScript-DCO-1.0-Signed-off-by: Zoltan Herczeg zherczeg@inf.u-szeged.hu

### CODE REVIEW

**Comment by egavrin** [comment]
Please add reference to #323

**Comment by egavrin** [comment]
Good to me.

**Comment by zherczeg** [comment]
What do you mean by adding a reference? Adding a comment in the function? E.g:
TODO ("Needs a proper upper case implementation. See issue #323");

File: jerry-core/lit/lit-char-helpers.h [hunk 1]
```diff
@@ -220,4 +220,13 @@ extern uint32_t lit_char_hex_to_int (ecma_char_t);
  */
 extern bool lit_char_is_word_char (ecma_char_t);
 
+/*
+ * Utility functions for uppercasing / lowercasing
+ */
+
+#define MAXIMUM_OTHERCASE_LENGTH (3)
```
Comment (ruben-ayrapetyan) [inline jerry-core/lit/lit-char-helpers.h]: @zherczeg, could you, please, add `LIT_` prefix and add describing comment for the definition?

File: jerry-core/lit/lit-char-helpers.cpp [hunk 2]
```diff
@@ -328,3 +328,71 @@ lit_char_is_word_char (ecma_char_t c) /**< code unit */
...
+/**
+ * Returns the lowercase character sequence of an ecma character.
+ *
+ * Note: output_buffer_p must be able to hold at least MAXIMUM_OTHERCASE_LENGTH characters.
+ *
+ * @return the length of the lowercase character sequence
+ *         which is always between 1 and MAXIMUM_OTHERCASE_LENGTH.
+ */
+lit_utf8_size_t
+lit_char_to_lower_case (ecma_char_t character, ecma_char_t *output_buffer_p)
```
Comment (ruben-ayrapetyan) [inline jerry-core/lit/lit-char-helpers.cpp]: Comments to arguments are missing.

File: jerry-core/lit/lit-char-helpers.cpp [hunk 3]
```diff
@@ -328,3 +328,71 @@ lit_char_is_word_char (ecma_char_t c) /**< code unit */
...
+/**
+ * Returns the uppercase character sequence of an ecma character.
+ *
+ * Note: output_buffer_p must be able to hold at least MAXIMUM_OTHERCASE_LENGTH characters.
+ *
+ * @return the length of the uppercase character sequence
+ *         which is always between 1 and MAXIMUM_OTHERCASE_LENGTH.
+ */
+lit_utf8_size_t
+lit_char_to_upper_case (ecma_char_t character, ecma_char_t *output_buffer_p)
```
Comment (ruben-ayrapetyan) [inline jerry-core/lit/lit-char-helpers.cpp]: Comments to arguments are missing.

File: jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp [hunk 4]
```diff
@@ -508,6 +509,157 @@ ecma_builtin_string_prototype_object_substring (ecma_value_t this_arg, /**< this
...
+  ECMA_TRY_CATCH (check_coercible_val,
+                  ecma_op_check_object_coercible (this_arg),
+                  ret_value);
+
+  /* 2. */
+  ECMA_TRY_CATCH (to_string_val,
+                  ecma_op_to_string (this_arg),
+                  ret_value);
+
+  JERRY_ASSERT (ecma_is_value_string (to_string_val));
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp]: The check would also be performed in `ecma_get_string_from_value`.

File: jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp [hunk 5]
```diff
@@ -508,6 +509,157 @@ ecma_builtin_string_prototype_object_substring (ecma_value_t this_arg, /**< this
...
 
 /**
+ * Helper function to convert a string to upper or lower case.
+ *
+ * @return completion value
+ *         Returned value must be freed with ecma_free_completion_value.
+ */
+static ecma_completion_value_t
+ecma_builtin_string_prototype_object_conversion_helper (ecma_value_t this_arg, /**< this argument */
+                                                        bool lower_case) /**< provideThis flag */
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp]: `provideThis` flag?

**Comment by egavrin** [comment]
@zherczeg I mean commit message:
```
Implement toLowerCase and toUpperCase built-in functions. 

Related issue: #323

JerryScript-DCO-1.0-Signed-off-by: Zoltan Herczeg zherczeg@inf.u-szeged.hu
```

**Comment by egavrin** [comment]
fyi
```
./jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp:519: error: line is longer than 120 characters
```

File: jerry-core/lit/lit-char-helpers.h [hunk 6]
```diff
@@ -220,4 +220,13 @@ extern uint32_t lit_char_hex_to_int (ecma_char_t);
  */
 extern bool lit_char_is_word_char (ecma_char_t);
 
+/*
+ * Utility functions for uppercasing / lowercasing
+ */
+
+#define LIT_MAXIMUM_OTHERCASE_LENGTH (3)
```
Comment (ruben-ayrapetyan) [inline jerry-core/lit/lit-char-helpers.h]: @zherczeg, what is the meaning of the value?

**Comment by zherczeg** [comment]
Thank you for the comments. I hope fixed all of them.

File: jerry-core/lit/lit-char-helpers.cpp [hunk 7]
```diff
@@ -328,3 +328,73 @@ lit_char_is_word_char (ecma_char_t c) /**< code unit */
           || (c >= LIT_CHAR_ASCII_DIGITS_BEGIN && c <= LIT_CHAR_ASCII_DIGITS_END)
           || c == LIT_CHAR_UNDERSCORE);
 } /* lit_char_is_word_char */
+
+/**
+ * Returns the lowercase character sequence of an ecma character.
+ *
+ * Note: output_buffer_p must be able to hold at least LIT_MAXIMUM_OTHER_CASE_LENGTH characters.
```
Comment (ruben-ayrapetyan) [inline jerry-core/lit/lit-char-helpers.cpp]: @zherczeg, could you, please, add arguments like `buffer_size` to the function, and check the condition with `JERRY_ASSERT`?

File: jerry-core/lit/lit-char-helpers.cpp [hunk 8]
```diff
@@ -328,3 +328,73 @@ lit_char_is_word_char (ecma_char_t c) /**< code unit */
...
+  }
+
+  output_buffer_p[0] = character;
+  return 1;
+} /* lit_char_to_lower_case */
+
+/**
+ * Returns the uppercase character sequence of an ecma character.
+ *
+ * Note: output_buffer_p must be able to hold at least LIT_MAXIMUM_OTHER_CASE_LENGTH characters.
```
Comment (ruben-ayrapetyan) [inline jerry-core/lit/lit-char-helpers.cpp]: @zherczeg, could you, please, also update the function likewise?

**Comment by ruben-ayrapetyan** [comment]
After adding `buffer_size`-like arguments (https://github.com/Samsung/jerryscript/pull/365#discussion-diff-34454948, https://github.com/Samsung/jerryscript/pull/365#discussion_r34454964) the changes would look good to me.

**Comment by ruben-ayrapetyan** [comment]
@zherczeg, thank you for update.

By the way, `__attr_unused____` seems to be unnecessary for me, as `JERRY_ASSERT` references variables in both debug and release modes. Are there any issues without the attribute?

**Comment by ruben-ayrapetyan** [comment]
Looks good to me

**Comment by egavrin** [comment]
`make push`

**Comment by zherczeg** [comment]
Landed in 69655f44561c6bfe43c60f73c4ea02091734d85e

**Pull request closed without merging**